listener.go

```diff
@@ -0,0 +1,409 @@
+// SPDX-FileCopyrightText: 2023 Comcast Cable Communications Management, LLC
+// SPDX-License-Identifier: Apache-2.0
+
+package listener
+
+import (
+	"bytes"
+	"context"
+	"crypto/hmac"
+	"encoding/hex"
+	"encoding/json"
+	"fmt"
+	"hash"
+	"io"
+	"net/http"
+	"strings"
+	"sync"
+	"time"
+
+	"github.com/xmidt-org/webhook-schema"
+	"go.uber.org/multierr"
+	"go.uber.org/zap"
+)
+
+const (
+	webpaHeader = "X-Webpa-Signature"
+	xmidtHeader = "Xmidt-Signature"
+)
+
+// Listener provides a way to register a webhook and validate the callbacks.
+// It can be configured to register the webhook at a given interval, as well as
+// it can also be configured to accept multiple secrets and hash algorithms.
+type Listener struct {
+	m                sync.RWMutex
+	registration     *webhook.Registration
+	registrationOpts []webhook.Option
+	interval         time.Duration
+	client           *http.Client
+	ctx              context.Context
+	upstreamCtx      context.Context
+	shutdown         context.CancelFunc
+	update           chan struct{}
+	getAuth          func() (string, error)
+	logger           *zap.Logger
+	metrics          *Measure
+	opts             []Option
+	body             []byte
+	acceptedSecrets  []string
+	hashPreferences  []string
+	hashes           map[string]func() hash.Hash
+}
+
+// Option is an interface that is used to configure the webhook listener.
+type Option interface {
+	fmt.Stringer
+	apply(*Listener) error
+}
+
+// New creates a new webhook listener with the given registration and options.
+func New(r *webhook.Registration, opts ...Option) (*Listener, error) {
+	l := Listener{
```

Comment:
It seems like the caller is going to have to read the code to figure out what all the options should be?  listener_test doesn't really make it clear, at least not for me.   Can we tie an example to the PR?